src/orion/algo/robo/ablr/ablr_model.py

```diff
@@ -0,0 +1,474 @@
+
+        self._predict_dist: Callable[[Tensor], Normal] | None = None
+
+        input_dims: int = self.feature_map.in_features
+        self.x_mean: Tensor
+        self.x_var: Tensor
+        self.register_buffer("x_mean", torch.zeros([input_dims]))
+        self.register_buffer("x_var", torch.ones([input_dims]))
+
+        self.y_mean: Tensor
+        self.y_var: Tensor
+        self.register_buffer("y_mean", torch.zeros(()))
+        self.register_buffer("y_var", torch.ones(()))
+
+    def forward(
+        self, x_train: Tensor, y_train: Tensor
+    ) -> tuple[Tensor, Callable[[Tensor], Normal]]:
+        """Return the training loss and a function that will give the
+        predictive distribution over y for a given (un-normalized) input x_test.
+
+        NOTE: Assumes that `x_train`, `y_train` and an eventual `x_test` are
+        NOT already rescaled/normalized!
+        """
+        # Change dtypes and devices if necessary.
+        x_train = x_train.type_as(self.x_mean)
+        y_train = y_train.type_as(self.y_mean)
+
+        x_train = self._normalize_x(x_train)
+        y_train = self._normalize_y(y_train)
+
+        assert x_train.shape[-1] == self.x_mean.shape[-1], (x_train.shape, self.x_mean)
+        neg_mll, predictive_distribution_fn = self.predictive_mean_and_variance(
+            x_train, y_train
+        )
+
+        def predictive_distribution(x_test: Tensor) -> Normal:
+            x_test = self._normalize_x(x_test)
+            x_test = torch.as_tensor(x_test).type_as(self.x_mean)
+            y_pred_dist = predictive_distribution_fn(x_test)  # type: ignore
+            y_pred_dist = self._unnormalize_y(y_pred_dist)
+            return y_pred_dist
+
+        return neg_mll, predictive_distribution
+
+    def predictive_mean_and_variance(
+        self, x_t: Tensor, y_t: Tensor
+    ) -> tuple[Tensor, Callable[[Tensor], Normal]]:
+        """Replicates the first part of section 3.1 of the ABLR paper.
+
+        Returns the 'learning criterion' (i.e. the loss term), and a function
+        that gives the predictive distribution (Normal) for a given (normalized)
+        test input.
+        """
+        assert len(x_t.shape) == 2
+        n = x_t.shape[0]
+        y_t = y_t.reshape([-1, 1])
+
+        # Feature map for the dataset of task t.
+        phi_t = self.feature_map(x_t.float())
+        d = phi_t.shape[-1]
+
+        r_t = self.beta / self.alpha
+
+        if n > d:
+            return self._n_greater_than_d(x_t=x_t, y_t=y_t, r_t=r_t, phi_t=phi_t)
+        else:
+            return self._n_smaller_than_d(x_t=x_t, y_t=y_t, r_t=r_t, phi_t=phi_t)
+
+    def _n_greater_than_d(
+        self,
+        x_t: Tensor,
+        y_t: Tensor,
+        r_t: Tensor,
+        phi_t: Tensor,
+    ):
+        n = x_t.shape[0]
+
+        k_t = r_t * phi_t.T @ phi_t
+        l_t = try_get_cholesky(k_t)
+        l_t_inv = try_get_cholesky_inverse(l_t)
```

Comment:
I think this gives k_t_inv rather than l_t_inv. It computes `inv = (l_t, l_t^T)^{-1}`. From what I understand, we would need to compute `l_t_inv = L_t^T K_t^{-1}`, this coming from the fact that `l_t_inv = l_t^T (l_t l_t^T)^{-1}` (https://en.wikipedia.org/wiki/Cholesky_decomposition#Matrix_inversion).

Reply:
Fixed in f04a2d6077473323e193cd8f1b09a853c5a7d9d7 (I hope)

Reply:
So it worked? Did it solve the negative variance issue with any chance? :crossed_fingers:

Reply:
I believe it did!